Changes to make ImageMeta constructor work

bhalonen opened the pull request:
Unsure on restrictions on metadata keys. Any comments? I don't like that there is a "String" requirement.

bhalonen committed:
Changes to make ImageMeta constructor work

Unsure on restrictions on metadata keys.

bhalonen commented:
No restrictions on the key type seems better.
```
mutable struct ImageMeta{T,N,A<:AbstractArray,P<:AbstractDict} <: AbstractArray{T,N}
    data::A
    properties::P

    function ImageMeta{T,N,A,P}(data::AbstractArray, properties::P) where {T,N,A,P}
        new{T,N,A,P}(data, properties)
    end
end
```

bhalonen commented:
references issue #40.

Tokazama commented:
The properties field has to be `Dict{String,Any}` or there becomes a problem of adding new properties of different types. I'm sure this could be worked around somehow, but if the properties field is restricted to specific types I'd think that eliminates the point of a property field.

timholy commented on src/ImageMetadata.jl:
```diff
@@ -32,15 +32,15 @@ export
 `props`), or with `ImageMeta(A, prop1=val1, prop2=val2, ...)`.
 """
-mutable struct ImageMeta{T,N,A<:AbstractArray,P<:AbstractDict{String,Any}} <: AbstractArray{T,N}
+mutable struct ImageMeta{T,N,A<:AbstractArray,P<:AbstractDict{String}} <: AbstractArray{T,N}
```
Keep the Any here since it's needed to add more properties

timholy commented on src/ImageMetadata.jl:
```diff
@@ -32,15 +32,15 @@ export
         new{T,N,A,P}(data, properties)
     end
 end
-ImageMeta(data::AbstractArray{T,N}, props::AbstractDict{String,Any}) where {T,N} = ImageMeta{T,N,typeof(data),typeof(props)}(data,props)
```
Keep this constructor but add a new one (or two :wink:)

timholy commented on src/ImageMetadata.jl:
```diff
@@ -32,15 +32,15 @@ export
     end
 end
-ImageMeta(data::AbstractArray{T,N}, props::AbstractDict{String,Any}) where {T,N} = ImageMeta{T,N,typeof(data),typeof(props)}(data,props)
+ImageMeta(data::AbstractArray{T,N}, props::AbstractDict{S}) where {T,N,S<:AbstractString} = ImageMeta{T,N,typeof(data),typeof(props)}(data,props)
```
I'd use
```julia
ImageMeta(data::AbstractArray, props::Dict{<:AbstractString}) = ImageMeta(data, convert(Dict{String,Any}, props))
ImageMeta(data::AbstractArray, props::AbstractDict) = throw(ArgumentError("properties must be an AbstractDict{String,Any}"))
```

bhalonen replied:
Ok, sounds good. Breaking change, but pretty easy to fix. (edit, not sure this is actually breaking)

timholy reviewed (commented):
Thanks for tackling this. Here are some suggested changes.

bhalonen committed:
Tim Holy suggested changes

timholy commented on src/ImageMetadata.jl:
```diff
@@ -41,7 +41,9 @@ mutable struct ImageMeta{T,N,A<:AbstractArray,P<:AbstractDict{String,Any}} <: Ab
     end
 end
 ImageMeta(data::AbstractArray{T,N}, props::AbstractDict{String,Any}) where {T,N} = ImageMeta{T,N,typeof(data),typeof(props)}(data,props)
+ImageMeta(data::AbstractArray, props::AbstractDict{<:AbstractString}) = ImageMeta(data, convert(Dict{String,Any}, props))
```
I suspect we should insist on `props::Dict{<:AbstractString}` here, rather than generalizing it to `AbstractDict`. If someone passes a `SortedDict`, this will silently un-sort it. I suspect if they went to the trouble to construct it as a `SortedDict` they'd be unhappy about that.

bhalonen replied:
If it isn't AbstractDict the constructor becomes recursive. Would you prefer that line 43 becomes
``` ImageMeta(data::AbstractArray{T,N}, props::Dict{String,Any}) where {T,N} = ImageMeta{T,N,typeof(data),typeof(props)}(data,props)```

timholy replied:
Oh, you're right. You could add another "non-copying" method
```julia
ImageMeta(data::AbstractArray{T,N}, props::Dict{String,Any}) where {T,N} = ImageMeta{T,N,typeof(data),typeof(props)}(data,props)
```
That will be more specific than the `Dict{<:AbstractString}` method.

But you can keep the `AbstractDict` one too. What's going on here is that the only `AbstractDict` we know about here is the `Dict` in Base, so we can confidently provide a fallback. For an arbitrary `AbstractDict` we keep it if it has the right type parameters, and otherwise force the user to fix it manually.

bhalonen replied:
Is there anyway to get from a ```TypeName``` to the type itself? We could use ```typeof(props).name``` to get the right type name to keep the other types of Dicts working automagically.

timholy replied:
Not generically. I could create a type like this:

```
DictWithNElements{N,K,V} <: AbstractDict{K,V}
```

What do you do with that `N` parameter in front of the others?

bhalonen replied:
I had this in mind. Maybe just too much. Nothing I need now.
```ImageMeta(data::AbstractArray, props::AbstractDict{<:AbstractString}) = ImageMeta(data, convert(type_from_typename(typeof(props).name){String,Any}, props))```

This could potentially make the SortedDict argument automatically convert.

I see that it is dangerous.

Tokazama replied:
Goodness! I'm sorry about this. I've been working on a custom dictionary for image metadata [here](https://github.com/Tokazama/ImageFormats.jl/blob/master/src/imageproperties.jl) that may be helpful. Unfortunately I've been spread pretty thin with my projects lately so it's not yet a finished work.

bhalonen replied:
Thanks for the help, learned some stuff.

timholy replied:
This is nice work, particularly for a first PR. Much appreciated.

bhalonen committed:
Type safety on conversion to Dict{String,Any}

Prevents other types of Dicts (such as SortedDict) from being automatically converted to Dicts
